### PR TITLE
Adds forceRestore flag to restore telemetry events

### LINKED ISSUE #9569: Add telemetry to understand when a restore happens due to a rebuild
Specifically, none of our metrics currently track it, but we should. 

cc @dominoFire

### PR DESCRIPTION
**Pull request by dominoFire**
## Bug

Fixes: https://github.com/NuGet/Home/issues/9569
Regression: No
* Last working version:   
* How are we preventing it in future:   

## Fix

Adds `forceRestore` flag parameter to a RestoreTelemetryEvent

## Testing/Validation

Tests Added: No
Reason for not adding tests:  Modified existing tests 
Validation:  Via Telemetry tool
- When you open a project in VS, a restore event is triggered  with the flag on false 
- When you click on Build > Rebuild Solution or Build > Rebuild Project, a restore event is triggered with the flag is set to true
- When you right-click in a project and select Build in a project with no assets an NuGet automatic restore is enabled, an event is triggered with the flag set on false.

The flag that indicates when a rebuild is requested is `VSSOLNBUILDUPDATEFLAGS.SBF_OPERATION_FORCE_UPDATE` as shown below:

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreBuildHandler.cs#L156

### CODE REVIEW

**Review by donnie-msft — APPROVED** [review]
I'm not crazy about the word "force" in this context, but it's kind of a nit.

**Comment by zivkan** [comment]
Do we have telemetry that tells us the "reason" for a restore?

These are the reasons I can think of:
* Project nomination
* explicit restore (customer right clicked solution, clicked restore)
* pre-build event

This forceRestore flag can only be set in a subset of the pre-build events, where the customer selected "rebuild", rather than "build".

If we have a field that tells us the difference between the 3 reasons above, I propose it would be better to add a 4th, instead of adding this new field.

**Comment by nkolev92** [comment]
> Do we have telemetry that tells us the "reason" for a restore?

RestoreOperationSource is it. 
* Explicit => Restore command
* Implicit => Nomination
* OnBuild => Pre Build event

> If we have a field that tells us the difference between the 3 reasons above, I propose it would be better to add a 4th, instead of adding this new field.

We can maybe add a 4th value, but keep in mind that the enum in question is used in more code than just telemetry (maybe we can just make the telemetry do a nicer thing and not change the rest of the code, but I have some arguments about that too). 

I wouldn't want the code to become `if (OperationSource == RestoreOperationSource.OnBuild || OperationSource == RestoreOperationSource.OnRebuild)`. 

Finally keep in mind that what is important for us is the fact that we can't do a no-op. 

Today the only gesture that can set a force flag is rebuild. That might change, but the idea really is to just understand whether a no-op was possible. 
We can get the frequency of build vs rebuild in other ways if we needed to. 

@dominoFire did suggest adding a 4th one in chat and this was my reply to him. 

*tldr;*

It's more valuable to reason about whether we have a force, whether a no-op is possible, more than reasoning about build vs rebuild today. 

I'd love to hear additional arguments :)

**Review by nkolev92 — APPROVED** [review]
Wait on a resolution on the discussion before merging.

**Comment by dominoFire** [comment]
> Do we have telemetry that tells us the "reason" for a restore?

Yes, you may refer to [RestoreOperationSource](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.VisualStudio.Common/SolutionRestore/RestoreOperationSource.cs).

> These are the reasons I can think of:
> 
> * Project nomination
> * explicit restore (customer right clicked solution, clicked restore)
> * pre-build event

The mapping is:

* Project nomination = `Implicit`
* explicit restore (customer right clicked solution, clicked restore) = `Explicit `
* pre-build event = `OnBuild`

> This forceRestore flag can only be set in a subset of the pre-build events, where the customer selected "rebuild", rather than "build".
> 
> If we have a field that tells us the difference between the 3 reasons above, I propose it would be better to add a 4th, instead of adding this new field.

Sounds reasonable to add a 4th value on the enumeration. That way, we avoid adding a new property to the telemetry event. I will think about it and write a solution in this PR.

**Comment by nkolev92** [comment]
> Sounds reasonable to add a 4th value on the enumeration. That way, we avoid adding a new property to the telemetry event. I will think about it and write a solution in this PR.


Make sure you refer to the above: https://github.com/NuGet/NuGet.Client/pull/3450#issuecomment-644899132

Also note: https://github.com/NuGet/Home/issues/6987. 

We are missing a gesture to reevaluate the lock, missing a gesture to reevaluate the floating versions. 

We can reason about the gesture, but we really want to reason about the impact primarily. 
We can *still* reason about the gesture as is, but if(more like when, verbal feedback has been rebuild running a force doesn't make much sense) at some point we change what rebuild does to restore, than the OnRebuild thing is pointless.

**Comment by dominoFire** [comment]
> > Sounds reasonable to add a 4th value on the enumeration. That way, we avoid adding a new property to the telemetry event. I will think about it and write a solution in this PR.
> 
> Make sure you refer to the above: [#3450 (comment)](https://github.com/NuGet/NuGet.Client/pull/3450#issuecomment-644899132)
> 
> Also note: [NuGet/Home#6987](https://github.com/NuGet/Home/issues/6987).
> 
> We are missing a gesture to reevaluate the lock, missing a gesture to reevaluate the floating versions.
> 
> We can reason about the gesture, but we really want to reason about the impact primarily.
> We can _still_ reason about the gesture as is, but if(more like when, verbal feedback has been rebuild running a force doesn't make much sense) at some point we change what rebuild does to restore, than the OnRebuild thing is pointless.

### Summary

- No more enum values because it is used in other places
- The main goal is to know if we can replace with a no-op instead of a full restore
- In all events, we will know if the restore resulted in an no-op if NoOpCount > 0
- Lock file and floating versions projects seems no changes in the no-op results. I will follow-up if I need to modify telemetry capture.
- Below is the table of observed cases:

RestoreSourceOperation   value | ForceRestore = True | ForceRestore = False
-- | -- | --
Implicit: Nomination, SlnRestoreMgr | ? | Open project.
OnBuild: VS Build events | Rebuild gestrue | Build gesture
Explicit: Restore gesture command | Don’t care.  | Don't care. 

That said, I think we can merge this PR as is, what do you think @zivkan ?

**Comment by nkolev92** [comment]
explicit restore allows no-op.

**Comment by nkolev92** [comment]
Refer to https://github.com/NuGet/Home/blob/dev/designs/VisualStudio-PartialRestoreOptimization.md#visual-studio-restore-gestures.
